### ISSUE TITLE
to_dataframe function is blocked when gunicorn is run with worker class gevent

### ISSUE THREAD

**Issue by gsroot** [googleapis/python-bigquery#697]
When I run my flask app with worker-class=gevent on gunicorn, the server blocks.

- gunicorn command
gunicorn app:app --workers=5 --worker-class=gevent --threads=5 --timeout=1800 --log-level=DEBUG

- Source code
```
query = '...'
query_job = bigquery_client.query(query)\
query_job.to_dataframe()# to_dataframe function where block occurs
```

- Source code where a block in the bigquery library occurs (lines 678 to 680 of the google/cloud/bigquery/_pandas_helpers.py file)

```
try:
    frame = worker_queue.get(timeout=_PROGRESS_INTERVAL) # this point
    yield frame
```

- Python library version
  - python=3.7.10
  - gunicorn=20.1.0
  - gevnet=21.1.2
  - eventlet=0.30.2
  - google-cloud-bigquery=2.20.0
  - google-cloud-bigquery-storage=2.4.0
  - google-cloud-core=1.6.0
  - pyarrow=4.0.0

The same happens when worker-class is an eventlet. It does not occur when worker-class is gthread or sync.
The block in _pandas_helpers.py is executed in the following syntax, is it a thread-related problem?

    with concurrent.futures.ThreadPoolExecutor(max_workers=total_streams) as pool:

Why do blocks happen?

**Comment by gsroot** [comment]
gevent/gevent#1797
gevent monkey patch의 영향 일까요? bigquery 모듈에서 해당 monkey patch 호환되도록 처리할 수 있습니까?

**Comment by plamut** [comment]
@gsroot I'm not familiar with inner workings of gevent, but it appears to be caused with how it interacts with `ThreadPoolExecutor`?

The line where the code gets stuck should not block for longer than `_PROGRESS_INTERVAL` seconds, but apparently the patch by gevent changes this for some reason. Did the [comment](https://github.com/gevent/gevent/issues/1797#issuecomment-859477319) on the related gevent issue help?

There might be a workaround, however. By default, results are downloaded using a streaming BigQuery Storage API, which streams table data in the background using a thread pool, but it is also possible to revert to the REST API:
```py
query_job.to_dataframe(create_bqstorage_client=False)
```

The REST API is slower, but might still perform well enough if the result set size is not excessive.

**Comment by gsroot** [comment]
@plamut Thank you. Your comment helped me roughly guess the cause of the problem, but I'm also not sure about the inner workings of gevent and it seems complicated, so I'll use create_bqstorage_client = False . Seems like a good alternative.

**Comment by iqbalaydrus** [comment]
@gsroot I ran into more or less the same issue as you (I was using bigquery-sqlalchemy). My fix was patching the grpc lib as mentioned here [githubissue](https://github.com/grpc/grpc/issues/4629)

```
import gevent.monkey

gevent.monkey.patch_all()  # noqa

# place it under the main gevent patch
import grpc.experimental.gevent

grpc.experimental.gevent.init_gevent()  # noqa
```

**Comment by gsroot** [comment]
@iqbalaydrus It seems key clue. Referring to the githubissue, It's probably a greean-thread related issue and it seems to have been resolved with the 1.11 release. If bigquery lib use grpc, it will work well, too. Thanks.
